### DISCUSSION TITLE
Hibernate Reactive Panache joining relations

### DISCUSSION THREAD

**Discussion by agreedSkiing** [quarkusio/quarkus#29461] [Q&A] [answered]
Hi. 
So we are running into a bit of a problem and we don't know what is best to do. We want to use Hibernate Reactive with Panache and the repository pattern to query the @ManyToOne relation to its @OneToMany relation and the other way around depending on which rest endpoint/service is called, but we are running into a problem when we try to load the relation with HQL with a query similar to 

> select distinct o from Order o left join fetch o.lineItems

[Simplified Hibernate Reactive with Panache#simplified-queries](https://quarkus.io/guides/hibernate-reactive-panache#simplified-queries).

Are we doing it wrong? Is there an another way like doing it “programmatically” (don’t know if it is the right term) or with a simple query? What is the best practice to do this with reactive hibernate and Panache?

[a comment by an automated account is not shown]

**Comment by DavideD** [comment, marked as answer]
> but we are running into a problem when we try to load the relation with HQL

Hi, what kind of problem?

**Reply by agreedSkiing** [reply]
One of my teammates that is trying to join the below entetis is getting this problem below
```json
{"details": "Error id 49df44a2-4ba8-4480-a375-397453f2a2c6-6, org.hibernate.LazyInitializationException: HR000056: Collection cannot be initialized: se.digg.sdk.addressbook.db.entity.Address.codes - Fetch the collection using 'Mutiny.fetch', 'Stage.fetch', or 'fetch join' in HQL"}
```
with the query code being
```java
 return find(
            "SELECT c FROM Item c LEFT JOIN FETCH c.crate WHERE c.crate.resourceIdentifier = :identifier",
            Parameters.with("identifier", identifier)
        ) 
```

## Mapped Super Class

```java
package com.company.project.common.entity;

import java.time.Instant;
import java.util.UUID;
import javax.persistence.Column;
import javax.persistence.GeneratedValue;
import javax.persistence.Id;
import javax.persistence.MappedSuperclass;
import javax.persistence.PrePersist;
import javax.persistence.PreUpdate;
import javax.persistence.Version;
import lombok.NoArgsConstructor;
import org.hibernate.annotations.GenericGenerator;

import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;

@NoArgsConstructor
@Getter
@Setter
@MappedSuperclass
public abstract class AbstractEntity {

    public AbstractEntity(String createdBy, String modifiedBy, int version) {
        this.createdBy = createdBy;
        this.modifiedBy = modifiedBy;
        this.version = version;
    }

    @Id
    @GeneratedValue(generator = "resource-uuid")
    @GenericGenerator(name = "resource-uuid", strategy = "org.hibernate.id.UUIDGenerator")
    @Column(name = "resource_id", nullable = false, updatable = false)
    UUID resourceIdentifier;

    @Column(name = "created_at", nullable = false, updatable = false)
    Instant createdAt;

    @Column(name = "modified_at")
    Instant modifiedAt;

    @Column(name = "created_by")
    String createdBy;

    @Column(name = "modified_by")
    String modifiedBy;

    @Version
    @Column(name = "version", nullable = false)
    int version; // For optimistic locking

    @PrePersist
    protected void onCreate() {
        createdAt = Instant.now();
        modifiedAt = null;
    }

    @PreUpdate
    protected void onModify() {
        modifiedAt = Instant.now();
    }
}
```

## Parent

```java
package com.company.project.logistics.db.entity;

import java.util.Set;
import javax.persistence.CascadeType;
import javax.persistence.Column;
import javax.persistence.Entity;
import javax.persistence.EnumType;
import javax.persistence.Enumerated;
import javax.persistence.FetchType;
import javax.persistence.OneToMany;
import javax.persistence.Table;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;
import com.company.project.common.entity.AbstractEntity;

@NoArgsConstructor
@Entity
@Getter
@Setter
@Table(name = "crate")
public class Crate extends AbstractEntity {

    public Crate(String createdBy, String modifiedBy, int version, String crateIdentifier,
            crateType crateType, crateStatus crateStatus, String crateVersion, String name,
            String alternateName, String description, String maintainer, String source, Set<item> items) {
        super(createdBy, modifiedBy, version);
        this.crateIdentifier = crateIdentifier;
        this.crateType = crateType;
        this.crateStatus = crateStatus;
        this.crateVersion = crateVersion;
        this.name = name;
        this.alternateName = alternateName;
        this.description = description;
        this.maintainer = maintainer;
        this.source = source;
        this.items = items;
    }

    @Column(name = "system_identifier", nullable = false)
    String crateIdentifier;

    @Enumerated(EnumType.STRING)
    @Column(name = "system_type")
    CrateType crateType;

    @Enumerated(EnumType.STRING)
    @Column(name = "system_status")
    CrateStatus crateStatus;

    @Column(name = "system_version")
    String crateVersion;

    @Column(name = "system_name")
    String name;

    @Column(name = "alternateName")
    String alternateName;

    @Column(name = "description")
    String description;

    @Column(name = "maintainer")
    String maintainer;

    @Column(name = "system_source")
    String source;

    @OneToMany(fetch = FetchType.EAGER, cascade = CascadeType.ALL, mappedBy = "crate")
    Set<Item> items;
}
```

## Enumclass 1

```java
package com.company.project.logistics.db.entity;

public enum CrateType {
    PLASTIC, WOODEN, METAL, BOTTLE, CONTAINER, PET
}
```

## Enumclass 2

```java
package com.company.project.logistics.db.entity;

public enum CrateStatus {
    FILLING, FULL, EMPTY, EMPTYING, TRANSPORTING, DELIVERED
}
```

## Child entity

```java
package com.company.project.logistics.db.entity;

import javax.persistence.Column;
import javax.persistence.Entity;
import javax.persistence.FetchType;
import javax.persistence.JoinColumn;
import javax.persistence.ManyToOne;
import javax.persistence.Table;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;
import com.company.project.common.entity.AbstractEntity;

@NoArgsConstructor
@Getter
@Setter
@Entity
@Table(name = "item")
public class Item extends AbstractEntity {

    public Item(String createdBy, String modifiedBy, int version, String itemIdentifier, String displayName,
            String originalText, String additionalInfo) {
        super(createdBy, modifiedBy, version);
        this.itemIdentifier = itemIdentifier;
        this.displayName = displayName;
        this.originalText = originalText;
        this.additionalInfo = additionalInfo;
    }

    @Column(name = "item")
    String itemIdentifier;

    @Column(name = "display_name")
    String displayName;

    @Column(name = "original_text")
    String originalText;

    @Column(name = "additional_info")
    String additionalInfo;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "crate_resource_id")
    Crate crate;
}
```

**Reply by DavideD** [reply]
You are fetching `c.create`  but the error is about `Address.codes`:

> HR000056: Collection cannot be initialized: se.digg.sdk.addressbook.db.entity.Address.codes

In Hibernate Reactive, [any lazy association must be explicitly load](https://staging.hibernate.org/reactive/documentation/1.1/reference/html_single/#_fetching_lazy_associations) (one way or another ) or you will see this error when trying to access it.

The recommended approach is to use a query, but you can still use `org.hibernate.reactive.mutiny.Mutiny.fetch`.

**Reply by agreedSkiing** [reply]
Sorry got multiple stacktraces from my teammates and pasted the wrong one, the one from c.create

**Reply by DavideD** [reply]
If you are not accessing this association in your code directly, my guess it's that it's happening when you are converting the result to an HTTP response. In that case, you need to make sure that the collection has been fetched beforehand or you can ignore that field during serialization.

**Reply by DavideD** [reply]
> Sorry got multiple stacktraces from my teammates and pasted the wrong one, the one from c.create

Ok, so, what's the actual error says?
Can you create a reproducer so that I can have a look?

**Reply by DavideD** [reply]
Maybe you can try this query instead:
```
SELECT i FROM Item i LEFT JOIN FETCH c.crate c WHERE c.resourceIdentifier = :identifier
```

**Reply by agreedSkiing** [reply]
The error says the same thing but instead `com.company.project.logistics.db.entity.Item.crate`, and the operation we are trying to do is just a simple crud. I think our biggest problem is that we are new to reactive hibernate and panache which is the biggest problem and just the link to the reactive hibernate documentation has helped a bit, but we are still lost in how to handle Uni<List<T>> and fetch all their associations

**Reply by DavideD** [reply]
At first glance, I don't see anything wrong.

I will have to write a test and have a look.

**Reply by agreedSkiing** [reply]
So yeah the query work, he found the issue with his code is that the mapper that is being used is trying to access the list in the parent object.

**Reply by agreedSkiing** [reply]
The my other teammate sent me his code and that code was based on eager fetching our relations beforehand which caused his problems and then not having a statement like .chain that I got working after doing some test. The code after reworking it to go from parent to child for the complete parent with all the children which was his goal with the previous query. 
```java
    public Uni<Crate> getByCreateIdWithItems(String identifier) {
        Uni<Crate> crate = find("crateIdentifier = :oid", Parameters.with("oid", identifier)).singleResult();
        crate.chain(x -> Mutiny.fetch(x.getItems()));
        return crate;
    }
```
I have mostly been the person in our team that hasn't had time to experiment with our new codebase and been more a support person (that’s why this post)
Also now we are thinking why use Panache when most of us are used to Hibernate and JPQL and why not just use Reactive Hibernate which is a new learning experience in itself.

**Reply by DavideD** [reply]
Glad you've found the issue.
You could also rewrite it like this:
```
    public Uni<Crate> getByCreateIdWithItems(String identifier) {
        return find("crateIdentifier = :oid", Parameters.with("oid", identifier))
            .singleResult()
            .call(x -> Mutiny.fetch(x.getItems()));
    }
```

Note that the fetch will run a separate query for loading the association, it might perform better to eagerly load the association

**Reply by agreedSkiing** [reply]
Did notice that but can't find another way to do it programmatically like the code i always like to look at https://learn.microsoft.com/en-us/ef/ef6/querying/related-data#eagerly-loading, but wait it might be possible with a criteria query if we use the plain reactive hibernate instead of panache

**Reply by DavideD** [reply]
Ah, wait, are you saying that it only works with `Mutiny.fetch` and doesn't when you use `fetch join` in a HQL?

This sounds like a bug in Hibernate Reactive. If that's the case, can you create an issue on the Hibernate Reactive issue tracker:
https://github.com/hibernate/hibernate-reactive/issues

I'm a bit busy now but I will get back to it.

> Also now we are thinking why use Panache when most of us are used to Hibernate and JPQL and why not just use Reactive Hibernate which is a new learning experience in itself.

Panache is an API on top of Hibernate Reactive. It simplify some operations using the entities by providing some convenient methods to run the most common operations.
It's not a strict requirement and you can inject the `Mutiny.SessionFactory`, if you prefer. Or, you can access the `Mutiny.Session` via `Panache.getSession()`.

That said, this error doesn't seem related to Panache and I think it would happen anyway .

**Reply by agreedSkiing** [reply]
@DavideD do you have any tips to join a manytomany relation in a .getResultList()/.list() with call or chain?

**Reply by agreedSkiing** [reply]
> Ah, wait, are you saying that it only works with `Mutiny.fetch` and doesn't when you use `fetch join` in a HQL?
> 
> This sounds like a bug in Hibernate Reactive. If that's the case, can you create an issue on the Hibernate Reactive issue tracker: https://github.com/hibernate/hibernate-reactive/issues
> 
> I'm a bit busy now but I will get back to it.
> 
> > Also now we are thinking why use Panache when most of us are used to Hibernate and JPQL and why not just use Reactive Hibernate which is a new learning experience in itself.
> 
> Panache is an API on top of Hibernate Reactive. It simplify some operations using the entities by providing some convenient methods to run the most common operations. It's not a strict requirement and you can inject the `Mutiny.SessionFactory`, if you prefer. Or, you can access the `Mutiny.Session` via `Panache.getSession()`.
> 
> That said, this error doesn't seem related to Panache and I think it would happen anyway .

No no, `fetch join` works perfectly fine our problem was a bad json mapping that made our child object call the parent object which had no transaction context when trying to load the set of children.

**Reply by DavideD** [reply]
> @DavideD do you have any tips to join a manytomany relation in a .getResultList()/.list() with call or chain?

Not sure, it depends on the specific use case.
